### ISSUE TITLE
register qualified keywords in workspace/document symbols 

### ISSUE THREAD

**Issue by kkharji** [clojure-lsp/clojure-lsp#662]
**Is your feature request related to a problem? Please describe.**

I'm starting to use integrant heavily and I find my self doing the following to have the key searchable in my workspace and buffer symbols:

```clj
(declare register)
(defmethod init-key ::register [_ {:keys [db]}]
  (fn [{:keys [body-params]}]
    (try
      (ok (db/append db body-params))
      (catch Exception e
        (err (ex-data e))))))
```

**Describe the solution you'd like**
::keys should be part of symbols? or maybe special hook for init-key. basically, I want to remove the extra declare statement as workaround

**Describe alternatives you've considered**
Use declare

**Comment by ericdallo** [comment]
You meant if `::register` should be a var-definition just like what we have for re-frame/reg-event-fx?

**Comment by tami5** [comment]
> You meant if `::register` should be a var-definition just like what we have for re-frame/reg-event-fx?

Yes, so I don't have to declare register.

**Comment by ericdallo** [comment]
Sorry, But I don't understand what advantage this will bring, what is your use case? what is the navigation flow you do that you think clojure-lsp can help?

**Comment by tami5** [comment]
> Sorry, But I don't understand what advantage this will bring, what is your use case? what is the navigation flow you do that you think clojure-lsp can help?

No problem, it helps with navigation mostly, i.e. jumping between symbols. In integrant you'd define a key with `init-key` then in a configuration file you would initialize that key and pass it around. e.g. `::register` > `:my.namespace/register {:db #ig/ref ...}`

In development, I'd need to jump between symbols and edit the defmethod body which work great but for defmethod and namespaced keys this is impossible. 

My current work around is to declare the key, as var above the init-key.

Clojure-lsp can somehow or maybe kondo can recognize init/key ::key as a function.

Does it make sense now?

**Comment by ericdallo** [comment]
I think so, but it sounds too specific for this case, maybe we can add some context specific data for [integrant](https://github.com/weavejester/integrant), but not sure what or where, maybe @borkdude could have some idea?

**Comment by borkdude** [comment]
You can already navigate to keywords?

**Comment by jwhitlark** [comment]
I believe Calva uses this functionality to show the outline for a file.  It would be nice if defmethods showed up, they don't currently.

**Comment by mainej** [comment]
@tami5 and @jwhitlark here's a mental model about how clj-kondo and clojure-lsp handle defmulti and defmethod...

`defmulti` **defines** a var. For example, `(defmulti x :dispatch-key)` defines `x`. Anywhere `x` is used you can use clojure-lsp's **`find-definition`** to navigate to the defmulti.

`defmethod` **implements** a var. For example `(defmethod x :a ...)` and `(defmethod x [:b 1] ...)` are both implementations of `x`. Anywhere `x` is used you can use clojure-lsp's **`find-implementations`** to show those defmethods.

The reason that there isn't a version that goes to _just one_ of the defmethods is because the ["dispatch-val"](https://clojuredocs.org/clojure.core/defmethod) of a defmethod doesn't have to be a keyword, even though it often is. For example, in the second defmethod above, the dispatch-val is a vector `[:b 1]`. These dispatch-vals can be arbitrarily large. In theory, you could teach clojure-lsp to match them, and infer that `(x {:dispatch-key |[:b 1]})` should go to the second defmethod, but I'm skeptical that would be helpful, since most of the time your code will actually look like `(x {:dispatch-key |local-holding-dynamically-constructed-dispatch-key})`, in which case clojure-lsp would have a hard time inferring which defmethod to go to, if it could do it at all.

@tami5 `declare` is an elegant workaround. It's a way of saying "this dispatch-val could be anything but is in fact an auto-resolved qualified keyword".  The drawback of `declare` is that you have to remember to include it. If you don't like using `declare`, perhaps you can use `find-implementations`. It will show you all the defmethods, which is more than you want, but perhaps your editor can help filter to the one you're interested in. Or alternatively, as @borkdude is suggesting, from the keyword you can use a third clojure-lsp feature **`find-references`**, which will show all of the occurrences of the keyword. The defmethod is one of those occurrences, so it will be included, though again this may show many more occurrences than you're interested in.

@jwhitlark I believe the feature you're talking about is called ["document-symbols"](https://clojure-lsp.io/features/#show-all-symbols-on-current-file). (There's a chance Calva doesn't use clojure-lsp's tools for this, in which case you'd have to refer this request to the Calva folks.) I rarely use that feature, so I'm not super familiar with it, but I believe the technical way to phrase what you're asking for is that you'd like document-symbols to include var-usages that are `:defmethod true`, in addition to the var-definitions it currently shows. There'd be an open question about what should show up in the outline... presumably you'd like to see the name of the multifn `x`, but also the dispatch-val `:a` or `[:b 1]`. I'm not sure what document-symbol's capabilities are. Can it include additional info like a dispatch-val? How would we handle really long or multi-line dispatch-vals?

**Comment by jwhitlark** [comment]
@mainej `document-symbols` sounds like the right feature, from your description.  @PEZ on the Calva team referred me to clojure-lsp, so I think it's safe to say it's handled via lsp, although I haven't confirmed usage of `document-symbols`  As far as what it would show, from my personal use case, it makes sense to have both the multifn name and the dispatch key, and long lines would just be my own problem, on the rare cases that it shows up.   What do you think are next steps on this?

**Comment by mainej** [comment]
@jwhitlark it wouldn't hurt to open a separate issue, with a reference back to this one. We can discuss details there. (And note to self... need to consider performance implications of finding dispatch-val, since it isn't available in clj-kondo's analysis.)

**Comment by jwhitlark** [comment]
I think #1016 closes this issue?

**Comment by ericdallo** [comment]
Yep, it should be fixed by #1016
